src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -544,59 +509,69 @@ where
 
-                    trace!(
-                        "timely-{worker_id} remap({id}) minted new bindings: \
-                        updates={:?} \
-                        source_upper={} \
-                        trace_upper={}",
-                        &remap_trace_batch.updates,
-                        source_upper.pretty(),
-                        remap_trace_batch.upper.pretty()
-                    );
-
-                    let cap = cap_set.delayed(cap_set.first().unwrap());
-                    remap_output.give_container(&cap, &mut remap_trace_batch.updates);
-
-                    // If the last remap trace closed the input, we no longer
-                    // need to (or can) advance the timestamper.
-                    if remap_trace_batch.upper.is_empty() {
-                        return;
-                    }
+        let mut prev_probe: Option<Probe<FromTime>> = None;
+        let timestamp_interval_ms: u64 = timestamp_interval
+            .as_millis()
+            .try_into()
+            .expect("huge duration");
 
-                    cap_set.downgrade(remap_trace_batch.upper);
+        while !cap_set.is_empty() {
+            // If we are reclocking to the latest offset then we only mint bindings after a
+            // successful probe. Otherwise we fall back to the earlier behavior where we just
+            // record the ingested frontier.
+            let (binding_ts, cur_source_upper) = if reclock_to_latest {
+                let new_probe = probed_upper
+                    .wait_for(|new_probe| match (&prev_probe, new_probe) {
+                        (None, Some(_)) => true,
+                        (Some(prev), Some(new)) => prev.probe_ts < new.probe_ts,
+                        _ => false,
+                    })
+                    .await
+                    .unwrap();
+                prev_probe = (*new_probe).clone();
+                let probe = prev_probe.clone().unwrap();
+                (probe.probe_ts, probe.upstream_frontier)
+            } else {
+                ticker.tick().await;
+                // We only proceed if the source upper frontier is not the minimum frontier. This
+                // makes it so the first binding corresponds to the snapshot of the source, and
+                // because the first binding always maps to the minimum *target* frontier we
+                // guarantee that the source will never appear empty.
+                let ingested_upper = ingested_upper
+                    .wait_for(|f| *f.frontier() != [FromTime::minimum()])
+                    .await
+                    .unwrap()
+                    .frontier()
+                    .to_owned();
+
+                let now = (now_fn)();
+                let mut binding_ts = now - now % timestamp_interval_ms;
+                if (now % timestamp_interval_ms) != 0 {
+                    binding_ts += timestamp_interval_ms;
+                }
```

Comment:
Can you explain this for me? Are we always intending to step forward by `timestamp_interval_ms` whenever we're not in the first interval?

Reply:
> whenever we're not in the first interval?
I didn't understand what you mean by "first interval" but I'll try to explain what happens here.

Given a `now()` value we want to convert it into a multiple of the `timestamp_interval` (having times that generally line up produces less work in compute when we do joins etc). We have three choices. We either ceil it and select the next interval multiple, or we floor it and select the previous interval multiple, or we round it and select the closest interval multiple. This piece of code implements the ceil strategy.